SpaceEngineers/Program.cs

```diff
@@ -1,78 +1,166 @@
-		// category under 'Visual C# Items' on the left hand side, and select
-		// 'Utility Class' in the main area. Name it in the box below, and
-		// press OK. This utility class will be merged in with your code when
-		// deploying your final script.
-		//
-		// You can also simply create a new utility class manually, you don't
-		// have to use the template if you don't want to. Just do so the first
-		// time to see what a utility class looks like.
-		// 
-		// Go to:
-		// https://github.com/malware-dev/MDK-SE/wiki/Quick-Introduction-to-Space-Engineers-Ingame-Scripts
-		//
-		// to learn more about ingame scripts.
+		private const string GroupToToggle = "Toggle On Dock";
+		private const string GroupToDisable = "Disable On Dock";
+		private const string GroupToEnable = "Enable On Undock";
+		private const string DockingConnectorKey = "[AutoToggle]";
+		private const string ReinitializeArgument = "initialize";
+		private const bool AutoScan = true;
+
+		/// code below here
+
+		private List<IMyFunctionalBlock> blocksToEnable = new List<IMyFunctionalBlock>();
+		private List<IMyGasTank> tanksToStockpile = new List<IMyGasTank>();
+		private List<IMyBatteryBlock> batteriesToRecharge = new List<IMyBatteryBlock>();
+
+		private List<IMyFunctionalBlock> blocksToDisable = new List<IMyFunctionalBlock>();
+		private List<IMyGasTank> tanksToUnstockpile = new List<IMyGasTank>();
+		private List<IMyBatteryBlock> batteriesToAuto = new List<IMyBatteryBlock>();
+
+		private IMyShipConnector connector;
+		private bool wasDockedOnLastRun = false;
 
 		public Program() {
-			// The constructor, called only once every session and
-			// always before any other method is called. Use it to
-			// initialize your script. 
-			//     
-			// The constructor is optional and can be removed if not
-			// needed.
-			// 
-			// It's recommended to set Runtime.UpdateFrequency 
-			// here, which will allow your script to run itself without a 
-			// timer block.
-		}
+			Runtime.UpdateFrequency |= UpdateFrequency.Update100;
 
-		public void Save() {
-			// Called when the program needs to save its state. Use
-			// this method to save your state to the Storage field
-			// or some other means. 
-			// 
-			// This method is optional and can be removed if not
-			// needed.
+			Initialize();
 		}
 
 		public void Main(string argument, UpdateType updateSource) {
-			// The main entry point of the script, invoked every time
-			// one of the programmable block's Run actions are invoked,
-			// or the script updates itself. The updateSource argument
-			// describes where the update came from. Be aware that the
-			// updateSource is a  bitfield  and might contain more than 
-			// one update type.
-			// 
-			// The method itself is required, but the arguments above
-			// can be removed if not needed.
+			if (
+				connector == null ||
+				(AutoScan && ((updateSource & UpdateType.Update100) != 0)) || // re-scan every 2-ish minutes if desired

```

Comment:
Are you sure this isn't every 2-ish seconds? i thought Update100 was every 100 ticks, with there being multiple per second.

Reply:
There are 60 ticks per second, so ideally this would trigger once every 10/6 second (1.666... seconds).

However, the game doesn't guarantee that 100 ticks means exactly 100 ticks. In the worst case, it can be almost double the requested wait (apparently less so for every-tick execution). So, "about every 2 seconds" is a reasonable human-scale approximation.